lib/df-detection.c

```diff
@@ -0,0 +1,116 @@
+#include "linux/df-detection.h"
+#include <linux/context_tracking.h>
+#include <linux/entry-common.h>
+#include <linux/kernel.h>
+#include <linux/livepatch.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+
+void add_address(const void *addr, unsigned long len)
+{
+	if (current->addresses == NULL || current->pairs == NULL)
+		return;
+	if (current->num_read >= current->sz &&
+	    reallocate_extra_memory(current->sz, DF_MAX_RECORDS)) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		detect_intersection();
+		current->num_read++;
+	}
+}
+
+void start_system_call()
+{
+	current->addresses = (struct df_address_range *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_address_range), GFP_KERNEL);
+	current->sz = current->addresses ? DF_INIT_SIZE : 0;
+	current->num_read = 0;
+	current->pairs = (struct df_pair *)kmalloc_array(
+	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
+	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+	current->df_index = 0;
+}
+void end_system_call()
```

Comment:
According to the kernel style guide, this should be written as `void end_system_call(void)`.